### PR TITLE
Slimmer #includes behaviour

### PR DESCRIPTION
**Pull request by ardecvz**
The goal of the commit is to make `#includes` slimmer - only join the table required to satisfy the referenced tables and preload all other associations independently.

### Motivation / Background

The Active Record #includes method results in different behaviour depending on the other parts of the query:
it may execute multiple queries (one per included association) or build a single JOIN query to load everything at once.

Before the commit, it was "all or nothing" behind an eager load flag - one referenced table was enough to transform the whole query into one big JOIN.

However, it's not always necessary to join all the tables except the ones with conditions.
Normally, we could avoid joining everything (which could be performance-heavy), and execute a few queries instead.

The following two statements would demonstrate the difference:

```sh
> Author.includes(:post, books: :essay).to_a

  Author Load (0.2ms)  SELECT "authors".* FROM "authors"
  Post Load (0.3ms)  SELECT "posts".* FROM "posts" WHERE "posts"."author_id" IN (?, ?, ?)  [["author_id", 1], ["author_id", 2], ["author_id", 3]]
  Book Load (0.3ms)  SELECT "books".* FROM "books" WHERE "books"."author_id" IN (?, ?, ?)  [["author_id", 1], ["author_id", 2], ["author_id", 3]]
  Essay Load (0.1ms)  SELECT "essays".* FROM "essays" WHERE "essays"."book_id" IN (?, ?, ?, ?)  [["book_id", 1], ["book_id", 3], ["book_id", 2], ["book_id", 4]]

> Author.includes(:post, books: :essay).where(books: { language: 'EN' }).to_a

  SQL (0.3ms)  SELECT "authors"."id" AS t0_r0, "authors"."name" AS t0_r1, "authors"."author_address_id" AS t0_r2, "authors"."author_address_extra_id" AS t0_r3, "authors"."organization_id" AS t0_r4, "authors"."owned_essay_id" AS t0_r5, "posts"."id" AS t1_r0, "posts"."author_id" AS t1_r1, "posts"."title" AS t1_r2, "posts"."body" AS t1_r3, "posts"."type" AS t1_r4, "posts"."legacy_comments_count" AS t1_r5, "posts"."taggings_with_delete_all_count" AS t1_r6, "posts"."taggings_with_destroy_count" AS t1_r7, "posts"."tags_count" AS t1_r8, "posts"."indestructible_tags_count" AS t1_r9, "posts"."tags_with_destroy_count" AS t1_r10, "posts"."tags_with_nullify_count" AS t1_r11, "books"."id" AS t2_r0, "books"."author_id" AS t2_r1, "books"."format" AS t2_r2, "books"."format_record_id" AS t2_r3, "books"."format_record_type" AS t2_r4, "books"."name" AS t2_r5, "books"."status" AS t2_r6, "books"."last_read" AS t2_r7, "books"."nullable_status" AS t2_r8, "books"."language" AS t2_r9, "books"."author_visibility" AS t2_r10, "books"."illustrator_visibility" AS t2_r11, "books"."font_size" AS t2_r12, "books"."difficulty" AS t2_r13, "books"."cover" AS t2_r14, "books"."isbn" AS t2_r15, "books"."external_id" AS t2_r16, "books"."original_name" AS t2_r17, "books"."published_on" AS t2_r18, "books"."boolean_status" AS t2_r19, "books"."tags_count" AS t2_r20, "books"."created_at" AS t2_r21, "books"."updated_at" AS t2_r22, "books"."updated_on" AS t2_r23, "essays"."id" AS t3_r0, "essays"."type" AS t3_r1, "essays"."name" AS t3_r2, "essays"."writer_id" AS t3_r3, "essays"."writer_type" AS t3_r4, "essays"."category_id" AS t3_r5, "essays"."author_id" AS t3_r6, "essays"."book_id" AS t3_r7 FROM "authors" LEFT OUTER JOIN "posts" ON "posts"."author_id" = "authors"."id" LEFT OUTER JOIN "books" ON "books"."author_id" = "authors"."id" LEFT OUTER JOIN "essays" ON "essays"."book_id" = "books"."id" WHERE "books"."language" IS NULL

```

### Detail

The main idea is to remove referenced tables
(which are #references, #where, #order or copied by #and, #or, #merge) from the final join query.
So we would LEFT OUTER JOINed only includes tables with conditions on the query. At the same time,
we have to bring non-referenced tables back to separate query preloads:

```sh
> Author.includes(:post, books: :essay).where(books: { language: 'EN' }).to_a

  SQL (0.2ms)  SELECT "authors"."id" AS t0_r0, "authors"."name" AS t0_r1, "authors"."author_address_id" AS t0_r2, "authors"."author_address_extra_id" AS t0_r3, "authors"."organization_id" AS t0_r4, "authors"."owned_essay_id" AS t0_r5, "books"."id" AS t1_r0, "books"."author_id" AS t1_r1, "books"."format" AS t1_r2, "books"."format_record_id" AS t1_r3, "books"."format_record_type" AS t1_r4, "books"."name" AS t1_r5, "books"."status" AS t1_r6, "books"."last_read" AS t1_r7, "books"."nullable_status" AS t1_r8, "books"."language" AS t1_r9, "books"."author_visibility" AS t1_r10, "books"."illustrator_visibility" AS t1_r11, "books"."font_size" AS t1_r12, "books"."difficulty" AS t1_r13, "books"."cover" AS t1_r14, "books"."isbn" AS t1_r15, "books"."external_id" AS t1_r16, "books"."original_name" AS t1_r17, "books"."published_on" AS t1_r18, "books"."boolean_status" AS t1_r19, "books"."tags_count" AS t1_r20, "books"."created_at" AS t1_r21, "books"."updated_at" AS t1_r22, "books"."updated_on" AS t1_r23 FROM "authors" LEFT OUTER JOIN "books" ON "books"."author_id" = "authors"."id" WHERE "books"."language" IS NULL
  Post Load (0.2ms)  SELECT "posts".* FROM "posts" WHERE "posts"."author_id" IN (?, ?, ?)  [["author_id", 1], ["author_id", 2], ["author_id", 3]]
  Essay Load (0.1ms)  SELECT "essays".* FROM "essays" WHERE "essays"."book_id" = ?  [["book_id", 1]]

```

### Backward compatibility

1) Includes and table names don't always match so we use reflections.

We try to match all possible includes table names against references including inner and left outer joins.
If there's an unreliable match (like no reflection for association), we're inclined to JOIN the associations in order to not break user applications.

All current tests are green without changes inside them (except a number of queries).

2) The commit certainly breaks user code with the plain string `where` or `select`:
```ruby
Author.includes(:company, :books).where(company: { title: "The Office" })
                                 .where("books.title = ?", "Design Patterns")
                                 .select("books.title")
```
We do not want to introduce any flags or scopes to support this behaviour because in docs there're mentions to prohibit string causes without references:
https://api.rubyonrails.org/v7.0.4.2/classes/ActiveRecord/QueryMethods.html#method-i-includes https://api.rubyonrails.org/v7.0.4.2/classes/ActiveRecord/QueryMethods.html#method-i-references

People who don't know it have to start using it.

### Known limitations

1) Includes may be ambiguous - a single table is enough to be referenced only once to remove the slim optimization from all `includes` trees.
Look at this example:
```
Author.includes(company: :authors, publishers: :authors).where(company: { authors: { name: "J. R. R. Tolkien" } })
```
Instead of the expected separate preload for the `publishers`, it would include them in a single JOIN. It's because we cannot differentiate usage level in the plain `references_values` at the moment.

Moreover, using references to numbered table aliases (with the digits at the end) disables all slim optimizations completely.

### Checklist

Before submitting the PR make sure the following are checked:

* [x] This Pull Request is related to one change. Changes that are unrelated should be opened in separate PRs.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [x] Tests are added or updated if you fix a bug or add a feature.
* [x] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

### CODE REVIEW

File: activerecord/lib/active_record/relation/calculations.rb [hunk 1]
```diff
@@ -189,7 +189,7 @@ def async_sum(identity_or_column = nil)
     #      end
     def calculate(operation, column_name)
       if has_include?(column_name)
-        relation = apply_join_dependency
+        relation = apply_join_dependency(full_eager_loading: true)
```
Comment (palkan) [inline activerecord/lib/active_record/relation/calculations.rb:192]: Why can't we use the same logic for calculations? I mean, why not taking referenced columns into account and ignore everything else?

File: activerecord/lib/active_record/relation/includes_tracker.rb [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+      def inferable_reflection_table_name?(reflection)
+        !reflection.through_reflection?
+      end
+
+      def join_table_with_postfix_alias(reflection)
+        [reflection.join_table, reflection.alias_candidate(:join)].sort.join("_")
+      end
+
+      def normalized_includes_values
+        includes_values.map do |element|
```
Comment (palkan) [inline activerecord/lib/active_record/relation/includes_tracker.rb]: Can't we use `#flat_map` here?

File: activerecord/lib/active_record/relation/includes_tracker.rb [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+        all_possible_includes_tables =
+          reflection.collect_join_chain.map(&:table_name) <<
+          reflection.alias_candidate(reflection.table_name) <<
+          reflection.name.to_s
+
+        if inferable_reflection_table_name?(reflection)
+          all_possible_includes_tables <<
+            reflection.join_table <<
+            join_table_with_postfix_alias(reflection)
+        end
```
Comment (palkan) [inline activerecord/lib/active_record/relation/includes_tracker.rb]: Can you add a comment or provide some details on why we need to take all these things into account? Like, there is `reflection.table_name`, why it's not enough?

File: activerecord/lib/active_record/relation/includes_tracker.rb [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+
+module ActiveRecord
+  # Determine includes which are used / not used in refererences and joins.
+  module IncludesTracker # :nodoc:
+    def includes_values_referenced
+      select_includes_values_with_references(:any?, :present?)
+    end
+
+    def includes_values_non_referenced
+      select_includes_values_with_references(:all?, :blank?)
```
Comment (palkan) [inline activerecord/lib/active_record/relation/includes_tracker.rb]: Passing predicates as parameters makes the code barely readable; looks like a de-duplication for the sake of de-duplication.

Maybe, we can calculate something once and re-use? Some intermediate value, which can be used to calculate both non-referenced and referenced.

Right now it's hard to follow the logic of building these values. It's okay to sacrifice _code metrics_ in favor of readability.

**Review by palkan — COMMENTED** [review]
Thanks!

Looks more sophisticated than I expected, but seems working. I left a few questions, let's discuss them and see how we can improve the solution.

File: activerecord/lib/active_record/relation/calculations.rb (hunk 1 above)
Comment (ardecvz) [reply activerecord/lib/active_record/relation/calculations.rb:192]: It's for backward compatibility - someone decided that it's nice to use includes as joins for `pluck` and `calculate`:
https://github.com/ardecvz/rails/pull/1/files#diff-6243fbad0c942a0aa315ce57c116984ce355f4e1742036a9de97c70f924a7745R167
Here `developer_id` is not from the parent table, it's from the joined one. It works without references in the current code. Yeah, really strange and crazy stuff.
I'm a big fan of Tests as documentation so I've transferred, improved, and named all these legacy things.
Also, I'm going to use another field in the example to communicate the idea better, thanks!

File: activerecord/lib/active_record/relation/includes_tracker.rb (hunk 2 above)
Comment (ardecvz) [reply activerecord/lib/active_record/relation/includes_tracker.rb]: The whole tree traversal is already rewritten to support an example from your Gist Readme (https://github.com/ardecvz/rails/pull/1/files#diff-6243fbad0c942a0aa315ce57c116984ce355f4e1742036a9de97c70f924a7745R65). 
It should be a simpler solution. 
Just two legacy broken tests to fix after the change. I'll push a preview and ping you again when it's ready.

File: activerecord/lib/active_record/relation/includes_tracker.rb (hunk 3 above)
Comment (ardecvz) [reply activerecord/lib/active_record/relation/includes_tracker.rb]: Again, lots of legacy support - I've documented them in the tests:
https://github.com/ardecvz/rails/pull/1/files#diff-6243fbad0c942a0aa315ce57c116984ce355f4e1742036a9de97c70f924a7745R210
I'll think how to explain it in the code.

File: activerecord/lib/active_record/relation/includes_tracker.rb (hunk 4 above)
Comment (ardecvz) [reply activerecord/lib/active_record/relation/includes_tracker.rb]: As I said before, a part of the code was rewritten to simplify code with recursion traversal.
Sadly, intersect_matcher is still here :'(
I was thinking about saving results to an intermediate variable but it seems to be against Active Record code around - it seemed to me that developers add state only when it's possible to reuse it (like `smth_values` which represents something real in the relation's life cycle).
I'm not sure to add some temporary variable to the relation but I'll try to figure out what to do with intersect_matcher, thanks!

File: activerecord/lib/active_record/relation/includes_tracker.rb (hunk 3 above)
Comment (ardecvz) [reply activerecord/lib/active_record/relation/includes_tracker.rb]: Explained with a comment.

File: activerecord/lib/active_record/relation/includes_tracker.rb [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+
+        if inferable_reflection_table_name?(reflection)
+          all_possible_includes_tables << reflection.join_table
+        end
+
+        all_possible_includes_tables
+      end
+
+      def references_table_with_postfix_alias?(all_references)
+        all_references.any? { _1.end_with?(TABLE_WITH_POSTFIX_ALIAS) }
```
Comment (skryukov) [inline activerecord/lib/active_record/relation/includes_tracker.rb]: This part looks rather hacky–what if I will call my table `foo_joins`? Is there a better way to check for postfix aliases? (btw, is that "Join table for HABTM" or smth else?)

Comment (ardecvz) [reply activerecord/lib/active_record/relation/includes_tracker.rb]: Thanks for the heads-up!
I've changed this part to the previous version of the code - generating a join table alias with the `_join` postfix.
Also, I've rephrased a comment to explain possible table names - yes, it's legacy? alias for the join table name (it's from the current Rails tests).
However, when I'd read the code I found a new interesting case - numbered table aliases:
```sql
  SQL (0.4ms)  SELECT "developers"."id" AS t0_r0, "developers"."name" AS t0_r1, "developers"."salary" AS t0_r2, "developers"."firm_id" AS t0_r3, "developers"."mentor_id" AS t0_r4, "developers"."legacy_created_at" AS t0_r5, "developers"."legacy_updated_at" AS t0_r6, "developers"."legacy_created_on" AS t0_r7, "developers"."legacy_updated_on" AS t0_r8, "projects"."id" AS t1_r0, "projects"."name" AS t1_r1, "projects"."type" AS t1_r2, "projects"."firm_id" AS t1_r3, "projects"."mentor_id" AS t1_r4, "developers_projects_2"."id" AS t2_r0, "developers_projects_2"."name" AS t2_r1, "developers_projects_2"."salary" AS t2_r2, "developers_projects_2"."firm_id" AS t2_r3, "developers_projects_2"."mentor_id" AS t2_r4, "developers_projects_2"."legacy_created_at" AS t2_r5, "developers_projects_2"."legacy_updated_at" AS t2_r6, "developers_projects_2"."legacy_created_on" AS t2_r7, "developers_projects_2"."legacy_updated_on" AS t2_r8, "projects_developers"."id" AS t3_r0, "projects_developers"."name" AS t3_r1, "projects_developers"."type" AS t3_r2, "projects_developers"."firm_id" AS t3_r3, "projects_developers"."mentor_id" AS t3_r4 FROM "developers" LEFT OUTER JOIN "developers_projects" ON "developers_projects"."developer_id" = "developers"."id" LEFT OUTER JOIN "projects" ON "projects"."id" = "developers_projects"."project_id" LEFT OUTER JOIN "developers_projects" "developers_projects_projects_join" ON "developers_projects_projects_join"."project_id" = "projects"."id" LEFT OUTER JOIN "developers" "developers_projects_2" ON "developers_projects_2"."id" = "developers_projects_projects_join"."developer_id" LEFT OUTER JOIN "projects" "projects_developers" ON "projects_developers"."id" = "developers_projects"."project_id" WHERE "developers_projects_2"."salary" IS NULL
```
And they can be referenced too:
https://github.com/ardecvz/rails/pull/1/files#diff-6243fbad0c942a0aa315ce57c116984ce355f4e1742036a9de97c70f924a7745R241
There're no tests in Rails for this behaviour (honestly, it's a little strange to filter something by automatic AR aliases), but I included code and tests to cover this in order to not break user applications (we'll cancel the optimization entirely for these strange cases).

File: activerecord/lib/active_record/relation/includes_tracker.rb (hunk 4 above)
Comment (ardecvz) [reply activerecord/lib/active_record/relation/includes_tracker.rb]: Besides a new tree traversal, I've changed intersect matchers for better readability.
You may read the methods almost as plain English.
While I'm thinking about adding an intermediate variable I prepared a quick benchmark - according to the results, the tree traversal costs almost nothing (it's within an error difference). So I think it's OK to traverse these small includes trees two times for a query at most (as a reward, we do not need to track changes to lots of variables used in includes or references calculations).
To check the benchmark I've also included a version to test separate preloads.
It's much slower than the old version : ) 
But it's expected as the query and the data are simple but allocating new objects is not.

Comment (ardecvz) [reply activerecord/lib/active_record/relation/includes_tracker.rb]: Benchmark results: https://github.com/ardecvz/rails/pull/2

Comment (ardecvz) [reply activerecord/lib/active_record/relation/includes_tracker.rb]: Also, I've updated README with known limitations - please, review it again.

Comment (ardecvz) [reply activerecord/lib/active_record/relation/includes_tracker.rb]: Added a honest case (preparing SQL) - we won around x0.7 at the database level (faster by 30%).

File: activerecord/test/cases/associations/nested_through_associations_test.rb [hunk 6]
```diff
@@ -539,7 +539,7 @@ def test_nested_has_many_through_with_conditions_on_through_associations
   def test_nested_has_many_through_with_conditions_on_through_associations_preload
     assert_empty Author.where("tags.id" => 100).joins(:misc_post_first_blue_tags)
 
-    author = assert_queries(2) { Author.includes(:misc_post_first_blue_tags).third }
+    author = assert_queries(3) { Author.includes(:misc_post_first_blue_tags).third }
```
Comment (palkan) [inline activerecord/test/cases/associations/nested_through_associations_test.rb:542]: Hm, this change looks suspicious; ideally, no existing tests should have changed. What's the source of the additional query?

Comment (ardecvz) [reply activerecord/test/cases/associations/nested_through_associations_test.rb:542]: I'm almost sure that it's a natural change - we're moving JOINs to separate preloads after all.
Before:
```sql
  FROM
  "posts"
  LEFT OUTER JOIN "taggings" ON "taggings"."taggable_type" = ?
  AND "taggings"."comment" = ?
  AND "taggings"."taggable_id" = "posts"."id"
  LEFT OUTER JOIN "tags" ON "tags"."name" = ?
  AND "tags"."id" = "taggings"."tag_id"
  LEFT OUTER JOIN "taggings" "taggings_tags_2" ON "taggings_tags_2"."tag_id" = "tags"."id"
```
After:
```sql
  -- The first query
  FROM
  "posts"
  LEFT OUTER JOIN "taggings" ON "taggings"."taggable_type" = ?
  AND "taggings"."comment" = ?
  AND "taggings"."taggable_id" = "posts"."id"
  LEFT OUTER JOIN "tags" ON "tags"."name" = ?
  AND "tags"."id" = "taggings"."tag_id"

  -- The second query
  FROM
  "taggings"
  LEFT OUTER JOIN "tags" ON "tags"."id" = "taggings"."tag_id"
  LEFT OUTER JOIN "taggings" "taggings_tags" ON "taggings_tags"."tag_id" = "tags"."id"
```
So we're preloading the second `taggings` instead of eager loading.
The change wasn't needed before because without the new tree traversal we preferred joining queries more frequently (but also we weren't able to optimize advanced things like [this](https://github.com/ardecvz/rails/pull/1/files#diff-6243fbad0c942a0aa315ce57c116984ce355f4e1742036a9de97c70f924a7745R65)).
By the way, we don't observe so many failures like this in the current Rails tests as there're not so many tests with both `includes` + no further `where` usage and a number of queries check.

Comment (ardecvz) [reply activerecord/test/cases/associations/nested_through_associations_test.rb:542]: However, it's nice that you've spotlighted this - it's strange that there's a JOIN `taggings` to `taggings` in the second query.
It's included in the long associations chain as the last `includes`:
```ruby
class Author < ActiveRecord::Base
  has_many :misc_post_first_blue_tags, through: :misc_posts, source: :first_blue_tags
  has_many :misc_posts, -> { where(posts: { title: ["misc post by bob", "misc post by mary"] }) }, class_name: "Post"
class Post < ActiveRecord::Base
  has_many :first_blue_tags, -> { where tags: { name: "Blue" } }, through: :first_taggings, source: :tag
  has_many :first_taggings, -> { where taggings: { comment: "first" } }, as: :taggable, class_name: "Tagging"
class Tagging < ActiveRecord::Base
  belongs_to :tag, -> { includes(:tagging) }
```
but it's still strange - I'm investigating. Maybe, it's because of the known limitations with ambiguous table names in `includes` and `references`.

Comment (ardecvz) [reply activerecord/test/cases/associations/nested_through_associations_test.rb:542]: Small update - if we remove `includes` in the last `belongs_to`, the problem would be solved.
It's super strange as we don't change how `includes` works - only filtering the current `includes_values`.
Right now, my plan is to replicate this behaviour without our optimization. It'd show this as another Rails minor issue.
